connectionpool.go

```diff
@@ -463,17 +479,34 @@ func (pool *hostConnPool) connectMany(count int) error {
 
 // create a new connection to the host and add it to the pool
 func (pool *hostConnPool) connect() (err error) {
+	return pool.connectToShard(-1)
+}
+
+func (pool *hostConnPool) connectToShard(shardHint int) (err error) {
 	// TODO: provide a more robust connection retry mechanism, we should also
 	// be able to detect hosts that come up by trying to connect to downed ones.
 	// try to connect
+
+	scParams := pool.scyllaHostInfo.makeConnParams(pool.session.connCfg, shardHint)
+	usesShardAwarePort := scParams.usesShardAwarePort()
+
 	var conn *Conn
 	reconnectionPolicy := pool.session.cfg.ReconnectionPolicy
 	for i := 0; i < reconnectionPolicy.GetMaxRetries(); i++ {
-		conn, err = pool.session.connect(pool.session.ctx, pool.host, pool)
+		conn, err = pool.session.connect(pool.session.ctx, pool.host, scParams, pool)
 		if err == nil {
 			break
 		}
 		if opErr, isOpErr := err.(*net.OpError); isOpErr {
+			// if we failed to bind, then probably it was due to the port being
+			// already allocated, so we can retry immediately with another port
+			if usesShardAwarePort && strings.Contains(opErr.Error(), "address already in use") {
+				if scParams.updateToNextPort() {
+					// don't consider it as a real trial
+					i--
+					continue
+				}
+			}
```

Comment:
Go does not seem to provide separate `bind` and `connect` operations. The `net.Dialer` allows specifying source address (including port), and has `Dial` function which both binds to the local address and tries to establish a connection. It's a little awkward for our case, as we'd like to know that a bind succeeded, and then tried to connect.

The problem is compounded that the fact that gocql also provides a `Dialer` interface which allows providing custom logic for dialing. I provided an extension `DialerExt` which accepts a source port. I don't really like the fact that we need to try multiple times in search of the port, each time calling custom client logic, but it seems to work.

Perhaps a different interface would be better? E.g. we would provide an iterator over a range of acceptable source ports, and the `DialerExt` would differentiate bind error from other errors internally?

Reply:
Apart from that, is it the right way to plug all of this into the retry logic?